File: TRT_MLP/MLP_Demo.cpp
/*
    ./mlp_demo -s ../mlp.wts ../mlp.engine
    ./mlp_demo -d ../mlp.wts ../mlp.engine
*/

#include <iostream>       
#include <map>
#include <string>         
#include <fstream>  
#include <chrono>  

#include "NvInfer.h"      
#include "logging.h"

using namespace nvinfer1;
static Logger gLogger;
const int INPUT_SIZE = 1;
const int OUTPUT_SIZE = 1;

class MLP_Demo{
public:
    MLP_Demo(std::string weight_path, std::string engine_path); // 构造函数
    void serialize(); // 序列化
    std::map<std::string, Weights> load_weight(const std::string file); // 加载权重文件
    // 创建engine
    ICudaEngine *createEngine(unsigned int maxBatchSize, IBuilder *builder, IBuilderConfig *config, DataType dt);
    // 推理
    void inference(float& input_data);
    void runInference(IExecutionContext &context, float *input, float *output, int batchSize);

private:
    std::string weight_path; // weight路径
    std::string engine_path; // engine路径
    unsigned int maxBatchSize = 1;
};

// 构造函数
MLP_Demo::MLP_Demo(std::string weight_path, std::string engine_path){
    this->weight_path = weight_path;
    this->engine_path = engine_path;
}

// 序列化
void MLP_Demo::serialize() {
    // Shared memory object
    IHostMemory *modelStream{nullptr};
    IBuilder *builder = createInferBuilder(gLogger); // 创建builder
    IBuilderConfig *config = builder->createBuilderConfig(); // 创建config
    ICudaEngine *engine = createEngine(this->maxBatchSize, builder, config, DataType::kFLOAT); // 创建engine
    assert(engine != nullptr);
    modelStream = engine->serialize(); // 以二进制序列化engine

    // 释放内存
    engine->destroy();
    builder->destroy();
    assert(modelStream != nullptr);

    std::ofstream p(this->engine_path, std::ios::binary); // 以二进制格式保存engine文件
    if(!p){
        std::cerr << "could not open plan output file" << std::endl;
        return;
    }
    p.write(reinterpret_cast<const char *>(modelStream->data()), modelStream->size());
    modelStream->destroy(); // 释放内存
}

// 加载权重文件
std::map<std::string, Weights> MLP_Demo::load_weight(const std::string file) {
    std::map<std::string, Weights> weightMap; // key: 权重名称 val: 权重数据
    std::ifstream input(file);
    assert(input.is_open() && "[ERROR]: Unable to load weight file...");

    int32_t count; // 权重数据的数量
    input >> count; // 2
    assert(count > 0 && "Invalid weight map file.");

    // 循环处理权重数据
    while(count--){
        Weights wt{DataType::kFLOAT, nullptr, 0}; // TensorRT的权重格式
        uint32_t size;
        std::string w_name;
        input >> w_name >> std::dec >> size;
        wt.type = DataType::kFLOAT;

        uint32_t *val = reinterpret_cast<uint32_t *>(malloc(sizeof(val) * size));
        for (uint32_t x = 0, y = size; x < y; ++x) {
            input >> std::hex >> val[x]; // 16进制（hex values）-> uint32
        }
        wt.values = val;
        wt.count = size;
        weightMap[w_name] = wt;
    }
    return weightMap;
}

ICudaEngine* MLP_Demo::createEngine(unsigned int maxBatchSize, IBuilder *builder, IBuilderConfig *config, DataType dt) {
    std::map<std::string, Weights> weightMap = load_weight(this->weight_path); // 读取权重文件
    INetworkDefinition *network = builder->createNetworkV2(0U); // 创建一个空的network
    ITensor *data = network->addInput("data", DataType::kFLOAT, Dims3{1, 1, 1}); // 创建engine的输入
    assert(data);

    // 在engine中创建一个线性层
    IFullyConnectedLayer *fc1 = network->addFullyConnected(*data, 1, weightMap["linear.weight"], weightMap["linear.bias"]);
    assert(fc1);
    fc1->getOutput(0)->setName("out"); // 设置fc1的输出名称
    network->markOutput(*fc1->getOutput(0)); // 将fc1层的输出标记为network的输出

    builder->setMaxBatchSize(1); // 设置最大batchsize
    config->setMaxWorkspaceSize(1 << 20); // 设置工作空间
    ICudaEngine *engine = builder->buildEngineWithConfig(*network, *config); // build engine
    assert(engine != nullptr);
    network->destroy(); // 释放network
    for (auto &mem: weightMap) { // 释放内存
        free((void *) (mem.second.values));
    }
    return engine; // 返回engine
}

// 反序列化加载engine并进行推理
void MLP_Demo::inference(float& input_data){
    char *trtModelStream{nullptr}; // 将模型写入 stream 中
    size_t size{0};

    std::ifstream file(this->engine_path, std::ios::binary); // 读取 engine
    if(file.good()){
        file.seekg(0, file.end);
        size = file.tellg();
        file.seekg(0, file.beg);
        trtModelStream = new char[size];
        assert(trtModelStream);
        file.read(trtModelStream, size);
        file.close();
    }

    IRuntime *runtime = createInferRuntime(gLogger); // 创建runtime
    assert(runtime != nullptr);

    ICudaEngine *engine = runtime->deserializeCudaEngine(trtModelStream, size, nullptr); // 反序列化加载engine
    assert(engine != nullptr);

    IExecutionContext *context = engine->createExecutionContext(); // 创建context
    assert(context != nullptr);

    float out[1]; // 初始化输出数组
    float data[1]; // 初始化输入数组
    for (float &i: data) i = input_data;

    auto start = std::chrono::system_clock::now(); // 记录推理时长

    runInference(*context, data, out, 1); // 执行推理

    auto end = std::chrono::system_clock::now(); // 记录推理时长
    std::cout << "\n[INFO]: Time taken by execution: "
              << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << "ms" << std::endl;

    // 释放内存和空间
    context->destroy();
    engine->destroy();
    runtime->destroy();

    // 打印输入和输出
    std::cout << "\nInput:\t" << data[0];
    std::cout << "\nOutput:\t";
    for (float i: out) {
        std::cout << i;
    }
    std::cout << std::endl;
}

void MLP_Demo::runInference(IExecutionContext &context, float *input, float *output, int batchSize) {
    const ICudaEngine &engine = context.getEngine(); // 从context中获取engine

    assert(engine.getNbBindings() == 2); // 输入和输出两个bindings
    void *buffers[2];

    // 通过name来获取输入和输出binding的索引
    const int inputIndex = engine.getBindingIndex("data");
    const int outputIndex = engine.getBindingIndex("out");

    // 在GPU中为输入和输出申请内存
    cudaMalloc(&buffers[inputIndex], batchSize * INPUT_SIZE * sizeof(float));
    cudaMalloc(&buffers[outputIndex], batchSize * OUTPUT_SIZE * sizeof(float));

    // 创建cudastream
    cudaStream_t stream;
    cudaStreamCreate(&stream);

    // 从 cpu 中拷贝输入到 gpu
    cudaMemcpyAsync(buffers[inputIndex], input, batchSize * INPUT_SIZE * sizeof(float), cudaMemcpyHostToDevice, stream);

    context.enqueue(batchSize, buffers, stream, nullptr); // 执行推理

    // 将推理结果从 gpu 拷贝回 cpu 中
    cudaMemcpyAsync(output, buffers[outputIndex], batchSize * OUTPUT_SIZE * sizeof(float), cudaMemcpyDeviceToHost, stream);

    cudaStreamSynchronize(stream); // 同步流，确保所有操作完成

    // 释放内存
    cudaStreamDestroy(stream);
    cudaFree(buffers[inputIndex]);
    cudaFree(buffers[outputIndex]);
}

void checkArgs(int argc, char **argv) {
    if (argc != 4) {
        std::cerr << "[ERROR]: Arguments not right!" << std::endl;
        std::cerr << "./mlp -s weight_path engine_path  // serialize model" << std::endl;
        std::cerr << "./mlp -d weight_path engine_path // deserialize and run inference" << std::endl;
        exit(1);
    }
}

int main(int argc, char **argv){
    checkArgs(argc, argv); // 检查参数
    MLP_Demo mlp_demo1(argv[2], argv[3]);

    if (std::string(argv[1]) == "-s") // 序列化
        mlp_demo1.serialize();
    else{ // 推理
        float input_test = 4.0;
        mlp_demo1.inference(input_test);
    }
    return 0;
}

File: TRT_MLP/MLP_Demo.py
'''
    python MLP_Demo.py -s --weight_path ./mlp.wts --engine_path ./mlp.engine
    python MLP_Demo.py -d --weight_path ./mlp.wts --engine_path ./mlp.engine
'''

import os
import struct
import argparse
import numpy as np
import pycuda.autoinit
import tensorrt as trt
import pycuda.driver as cuda

# 日志
gLogger = trt.Logger(trt.Logger.INFO)

class MLP_Demo():
    def __init__(self, weight_path, engine_path):
        super(MLP_Demo, self).__init__()
        self.weight_path = weight_path # weight路径
        self.engine_path = engine_path # engine路径
        self.input_blob_name = 'data' # 输入名称
        self.out_blob_name = 'out' # 输出名称
        self.input_size = 1
        self.output_size = 1
        
    # 序列化模型
    def serialize_model(self, max_batch_size):
        builder = trt.Builder(gLogger) # 创建builder
        config = builder.create_builder_config() # 创建config
        engine = self.create_engine(max_batch_size, builder, config, trt.float32) # 序列化模型
        assert engine

        with open(self.engine_path, "wb") as f: # 保存序列化好的 engine
            f.write(engine.serialize())
            
        # 释放内存
        del engine
        del builder

    def create_engine(self, max_batch_size, builder, config, dt):
        weight_map = self.load_weights(self.weight_path) # 加载模型
        network = builder.create_network() # 先使用 builder 创建一个空的network 
        data = network.add_input(self.input_blob_name, dt, (1, 1, self.input_size))  # 添加输入操作数
        assert data

        # 使用模型的权重来添加线性层
        linear = network.add_fully_connected(input = data, num_outputs = self.output_size,
                                            kernel = weight_map['linear.weight'],
                                            bias = weight_map['linear.bias'])
        assert linear
        
        linear.get_output(0).name = self.out_blob_name # 设置线性层的输出名称
        network.mark_output(linear.get_output(0)) # 将线性层的输出标记为整个网络的输出
        builder.max_batch_size = max_batch_size # 设置最大 batchsize
        engine = builder.build_engine(network, config) # 使用空的network和相应的配置config来创建engine

        # 释放内存
        del network
        del weight_map

        return engine # 返回创建好的engine
    
    # 加载权重数据，以 dict 的形式存放
    def load_weights(self, file_path):
        assert os.path.exists(file_path), '[ERROR]: Unable to load weight file.'

        weight_map = {}
        with open(file_path, "r") as f:
            lines = [line.strip() for line in f]

        count = int(lines[0])
        assert count == len(lines) - 1

        # 循环遍历处理权重数据
        for i in range(1, count + 1):
            splits = lines[i].split(" ")
            name = splits[0]
            cur_count = int(splits[1])
            assert cur_count + 2 == len(splits)

            values = []
            for j in range(2, len(splits)): # 循环遍历处理权重数据
                # bytes.fromhex 将16个字节的字符串转化为字节对象
                # struct.unpack 对字节对象进行解包
                values.append(struct.unpack(">f", bytes.fromhex(splits[j])))

            # 保存格式为 { 'weight.name': [weights_val0, weight_val1, ..] }
            weight_map[name] = np.array(values, dtype=np.float32)
        return weight_map
    
    
    def inference(self, inf_context, inf_host_in, inf_host_out):
        inference_engine = inf_context.engine # 推理引擎
        assert inference_engine.num_bindings == 2 # 输入和输出两个bindings
        device_in = cuda.mem_alloc(inf_host_in.nbytes) # 在GPU中分配输入的内存
        device_out = cuda.mem_alloc(inf_host_out.nbytes) # 在GPU中分配输出的内存

        bindings = [int(device_in), int(device_out)] # 创建输入和输出两个 bindings
        stream = cuda.Stream() # 创建cuda流
        cuda.memcpy_htod_async(device_in, inf_host_in, stream) # 将输入数据从cpu拷贝到gpu中

        # 执行推理
        inf_context.execute_async(bindings=bindings, stream_handle=stream.handle)

        cuda.memcpy_dtoh_async(inf_host_out, device_out, stream) # 将推理结果从gpu拷贝到cpu中
        stream.synchronize() # 同步流，确保所有操作完成
        
    # 反序列化加载序列化的engine，并进行模型推理
    def deserialize_inference(self, input_val):
        runtime = trt.Runtime(gLogger) # 创建runtime
        assert runtime
        
        # 反序列化加载模型，生成engine
        with open(self.engine_path, "rb") as f:
            engine = runtime.deserialize_cuda_engine(f.read())
        assert engine

        # 使用 engine 生成上下文
        context = engine.create_execution_context()
        assert context

        data = np.array([input_val], dtype=np.float32) # 创建输入数据
        host_in = cuda.pagelocked_empty((self.input_size), dtype=np.float32) # 在GPU申请输入的内存缓冲区
        np.copyto(host_in, data.ravel()) # 将输入数据拷贝到内存缓冲区
        host_out = cuda.pagelocked_empty(self.output_size, dtype=np.float32) # 在GPU申请输出的内存缓冲区

        # 执行推理
        self.inference(context, host_in, host_out) # 传入 context 和对应的输入输出缓冲区

        print(f'Input:\t{input_val}\nOutput:\t{host_out[0]:.4f}')

def get_parser():
    arg_parser = argparse.ArgumentParser(description = 'Parameters of MLP Demo')
    arg_parser.add_argument('--weight_path', # 模型的权重数据
                            type = str,
                            default = "./mlp.wts") # engine的保存地址
    arg_parser.add_argument('--engine_path', 
                            type = str,
                            default = "./mlp.engine")
    arg_parser.add_argument('-s', action = 'store_true') # 序列化模型
    arg_parser.add_argument('-d', action = 'store_true') # 推理模型
    return arg_parser

if __name__ == "__main__":
    parser = get_parser()
    args = parser.parse_args()
    mlp_demo = MLP_Demo(args.weight_path, args.engine_path) # 实例化demo对象
    if args.s: 
        mlp_demo.serialize_model(max_batch_size = 1) # 序列化模型
    else: # 反序列化模型进行推理
        mlp_demo.deserialize_inference(input_val = 4.0)

